oeps/oep-0031-arch-i18n.rst

```diff
@@ -0,0 +1,122 @@
+==============================
+OEP-0031: Internationalization
+==============================
+
+.. list-table::
+
+   * - OEP
+     - :doc:`OEP-0031 <oep-0031-arch-i18n>`
+   * - Title
+     - Internationalization
+   * - Last Modified
+     - 2019-01-07
+   * - Authors
+     -
+   * - Arbiter
+     -
+   * - Status
+     - Draft
+   * - Type
+     - Architecture
+   * - Created
+     - 2019-01-07
+   * - Review Period
+     -
+
+.. contents::
+   :local:
+   :depth: 2
+
+Motivation
+==========
+
+Internationalization (i18n) is a vital part of unlocking content for users around the globe and effectively supporting our business partners. In order to help us reuse localized components from app to app, the different parts of the edX codebase should align on a consistent handling of i18n. This document covers our approach to internationalizing new front-end applications and components written in React.
+
+In this OEP we use the terms locale and language more or less interchangeably. However, note that this document addresses primarily making code multilingual, i.e., the issue of presenting the same content in different languages. Considering ways to make it multi-regional, to allow for presenting different content for different countries, is out of scope for this OEP.
+
+Technology Choice
+=================
+
+We will use `react-intl <https://github.com/yahoo/react-intl>`_ to internationalize React components, in agreement with `Paragon ADR-2 <https://github.com/edx/paragon/blob/f5b9f876800e5154ef01d04f682b8cda21b3e439/docs/decisions/0002-react-i18n.rst>`_. react-intl is in line with current industry standards in i18n and satisfies all currently identified i18n needs, including multiple plurals.
+
+Alternatives Considered
+-----------------------
+
+An alternative choice we considered was `i18next <https://react.i18next.com/>`_. **[TODO]**
+
+Locale
```

Comment:
Have we decided on what should happen when a user changes their language preference?  Whether and how does that change get propagated to other browsers and frontends?

Reply:
Do you mean from a language switcher widget, as opposed to changing it permanently in their user settings?

Reply:
No, I didn't mean to suggest introducing a new way for users to change their language settings.  Question is what the behavior is cross-browser(-tab) and frontend when a user changes their language setting.  The answer to this may indicate requirements for invalidating caches/cookies/etc.

Reply:
@nasthagiri There are multiple ways user can change their language preference. Are you talking about user changing their language preference in Browser? Or user changes their language preference from their user profile edX page? Or a combination of them? If we call out the way we recognize the locale in a prioritized list, then we can infer the possibilities when user switch their language preference at certain places.

Reply:
@schenedx Good point.  For this particular design question, I was asking about when a user changes their persisted language preference in their edX account.  They may change it in a dropdown in the edX footer/header, in their profile page, or whatever other UX we support in the future.

Reply:
@nasthagiri Is this still a concern for you for this version of this OEP?  The current behavior is that other open tabs will just not see the change until the next page refresh, and to my knowledge changing that is not a high priority.